### PR TITLE
Handle errors better

### LINKED ISSUE #25: Handle errors better
Downloading all data for 2 digit HUC codes threw the following errors. Seems like a connection problem.  We should consider retries in these cases?

Implement design in #1

## ChunkEncodingError
```
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 696, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 436, in _error_catcher
    yield
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 763, in read_chunked
    self._update_chunk_length()
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 700, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 753, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 571, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 792, in read_chunked
    self._original_response.close()
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 454, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-93e40921627b>", line 1, in <module>
    output_h5.download_all('./data', hucs=hucs)
  File "/Users/val/workspace/stream-temperature-ml/playground/output_h5.py", line 39, in download_all
    download_data(f"{expanded_dir}/{filename}", **query_info)
  File "/Users/val/workspace/stream-temperature-ml/playground/output_h5.py", line 66, in download_data
    data_df, metadata, _, data_variables, nodata_variables = synthesis.get_timeseries_data(synthesizer, **query_info)
  File "/Users/val/workspace/basin3d/basin3d/synthesis.py", line 382, in get_timeseries_data
    for data_obj in data_generator:
  File "/Users/val/workspace/basin3d/basin3d/core/synthesis.py", line 146, in list
    for obj in plugin_views[self.synthesis_model].list(**synthesized_query_params):
  File "/Users/val/workspace/basin3d/basin3d/plugins/usgs.py", line 595, in list
    for data_json in generator_usgs_measurement_timeseries_tvp_observation(self, **kwargs):
  File "/Users/val/workspace/basin3d/basin3d/plugins/usgs.py", line 144, in generator_usgs_measurement_timeseries_tvp_observation
    params=search_params)
  File "/Users/val/workspace/basin3d/basin3d/core/access.py", line 33, in get_url
    response = requests.get(url, params=params, verify=verify, headers=headers, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/sessions.py", line 697, in send
    r.content
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 831, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 756, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))

### USGS error (more information needed)
USGS-05
```
ERROR:basin3d.plugins.usgs:could not convert string to float: 
```


### ContentDecodingError
USGS-15 - failed
```
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 399, in _decode
    data = self._decoder.decompress(data)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 86, in decompress
    ret += self._obj.decompress(data)
zlib.error: Error -3 while decompressing data: invalid block type
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 753, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 571, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 768, in read_chunked
    chunk, decode_content=decode_content, flush_decoder=False
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/urllib3/response.py", line 405, in _decode
    e,
urllib3.exceptions.DecodeError: ('Received response with content-encoding: gzip, but failed to decode it.', error('Error -3 while decompressing data: invalid block type'))
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-16-f8340aa6b187>", line 3, in <module>
    output_h5.download_all("./data", hucs=[h])
  File "/Users/val/workspace/stream-temperature-ml/playground/output_h5.py", line 39, in download_all
    download_data(f"{expanded_dir}/{filename}", **query_info)
  File "/Users/val/workspace/stream-temperature-ml/playground/output_h5.py", line 66, in download_data
    data_df, metadata, _, data_variables, nodata_variables = synthesis.get_timeseries_data(synthesizer, **query_info)
  File "/Users/val/workspace/basin3d/basin3d/synthesis.py", line 382, in get_timeseries_data
    for data_obj in data_generator:
  File "/Users/val/workspace/basin3d/basin3d/core/synthesis.py", line 146, in list
    for obj in plugin_views[self.synthesis_model].list(**synthesized_query_params):
  File "/Users/val/workspace/basin3d/basin3d/plugins/usgs.py", line 595, in list
    for data_json in generator_usgs_measurement_timeseries_tvp_observation(self, **kwargs):
  File "/Users/val/workspace/basin3d/basin3d/plugins/usgs.py", line 144, in generator_usgs_measurement_timeseries_tvp_observation
    params=search_params)
  File "/Users/val/workspace/basin3d/basin3d/core/access.py", line 33, in get_url
    response = requests.get(url, params=params, verify=verify, headers=headers, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/sessions.py", line 697, in send
    r.content
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 831, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
  File "/Users/val/anaconda3/envs/basin3d/lib/python3.7/site-packages/requests/models.py", line 758, in generate
    raise ContentDecodingError(e)
requests.exceptions.ContentDecodingError: ('Received response with content-encoding: gzip, but failed to decode it.', error('Error -3 while decompressing data: invalid block type'))
```

### PR DESCRIPTION
**Pull request by vchendrix**
Implements infrastructure for handling error messages from Datasource
plugings. See #1  for design.

Also adds the entire API Reference to the documentation. Small edits to code docstrings
due to autodoc warnings.

Closes #25

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Unit tests
- [X] Integration tests 
- [X] Test coverage >= 93%
- [X] Flake8 Tests 
- [X] Mypy Tests 

### Test Configuration
* Python Version: 3.7.5

## PR Self Evaluation
strikethrough things that don’t make sense for your PR

- [X] My code follows the agreed upon best practices
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests or modified existing tests that prove my fix is effective or that my feature works
- [X] Existing unit tests pass locally with my changes
- [X] I have performed a self-review of my own code

### CODE REVIEW

File: tests/test_synthesis.py [hunk 1]
```diff
@@ -10,19 +10,22 @@
...
     """Test basic plugin registration"""
 
-    from tests.testplugins import alpha
-    from basin3d.plugins import usgs
-    synthesizer = register()
+    synthesizer = register(["basin3d.plugins.usgs.USGSDataSourcePlugin",
+                            "tests.testplugins.alpha.AlphaSourcePlugin"])
 
     datasources = synthesizer.datasources
+    print(datasources)
```
Comment (dschristianson) [inline tests/test_synthesis.py:24]: remove?

File: tests/test_plugins_usgs.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def test_measurement_timeseries_tvp_observations_usgs(monkeypatch):
 def test_usgs_monitoring_feature(query, feature_type, monkeypatch):
     """Test USGS search by region  """
 
-    def mock_get_huc_codes(*args):
+    def mock_get_huc_codes(*args, **kwargs):
```
Comment (dschristianson) [inline tests/test_plugins_usgs.py:110]: why is kwargs needed here?

File: basin3d/plugins/usgs.py [hunk 3]
```diff
@@ -128,6 +128,7 @@ def generator_usgs_measurement_timeseries_tvp_observation(view,
         for stat in query.statistic:
             sythesized_stat = USGS_STATISTIC_MAP.get(stat)
             if not sythesized_stat:
+                synthesis_messages.append(f"USGS Daily Values service does not support statistic {stat}")
                 logger.info(f"USGS Daily Values service does not support statistic {stat}")
```
Comment (dschristianson) [inline basin3d/plugins/usgs.py:132]: Do we anticipate that the messages will be different between the synthesis_messages and logger?
If not, I suggest having a single message. Something like:
```
msg = f"USGS Daily Values service does not support statistic {stat}"
synthesis_message.append(msg)
logger.info(msg)
```

**Review by dschristianson — CHANGES_REQUESTED** [review]
Looks great, Val. Few minor suggestions.

File: tests/test_synthesis.py (hunk 1 above)
Comment (vchendrix) [reply tests/test_synthesis.py:24]: Agreed.

File: basin3d/plugins/usgs.py (hunk 3 above)
Comment (vchendrix) [reply basin3d/plugins/usgs.py:132]: @dschristianson you bring up a good point. My thought was to leave the logging messages in place for when we implement logger handling. I think this still needs to be thought through.

**Review by dschristianson — APPROVED** [review]
Looks great, @vchendrix.
Your responses make perfect sense.

**Pull request merged**